src/Features/Core/Portable/Diagnostics/CodeAnalysisDiagnosticAnalyzerService.cs

```diff
@@ -0,0 +1,115 @@
+        {
+            private readonly IDiagnosticAnalyzerService _diagnosticAnalyzerService;
+            private readonly IDiagnosticsRefresher _diagnosticsRefresher;
+            private readonly Workspace _workspace;
+            private readonly ConcurrentSet<ProjectId> _analyzedProjectIds = new();
+
+            public CodeAnalysisDiagnosticAnalyzerService(
+                IDiagnosticAnalyzerService diagnosticAnalyzerService,
+                IDiagnosticsRefresher diagnosticsRefresher,
+                Workspace workspace)
+            {
+                _diagnosticAnalyzerService = diagnosticAnalyzerService;
+                _diagnosticsRefresher = diagnosticsRefresher;
+                _workspace = workspace;
+
+                _workspace.WorkspaceChanged += OnWorkspaceChanged;
+            }
+
+            private void OnWorkspaceChanged(object? sender, WorkspaceChangeEventArgs e)
+            {
+                switch (e.Kind)
+                {
+                    case WorkspaceChangeKind.SolutionAdded:
+                    case WorkspaceChangeKind.SolutionCleared:
+                    case WorkspaceChangeKind.SolutionReloaded:
+                    case WorkspaceChangeKind.SolutionRemoved:
+                        _analyzedProjectIds.Clear();
+                        break;
+                }
+            }
+
+            public bool HasProjectBeenAnalyzed(ProjectId projectId) => _analyzedProjectIds.Contains(projectId);
+
+            public async Task RunAnalysisAsync(Solution solution, Action<Project> onProjectAnalyzed, ProjectId? projectId, CancellationToken cancellationToken)
+            {
+                if (projectId != null)
+                {
+                    var project = solution.GetProject(projectId);
+                    if (project != null)
+                    {
+                        await AnalyzeProjectCoreAsync(project, onProjectAnalyzed, cancellationToken).ConfigureAwait(false);
+                    }
+                }
+                else
+                {
+                    var tasks = new Task[solution.ProjectIds.Count];
+                    var index = 0;
+                    foreach (var project in solution.Projects)
+                    {
+                        tasks[index++] = Task.Run(() => AnalyzeProjectCoreAsync(project, onProjectAnalyzed, cancellationToken), cancellationToken);
+                    }
+
+                    await Task.WhenAll(tasks).ConfigureAwait(false);
+                }
+            }
+
+            private async Task AnalyzeProjectCoreAsync(Project project, Action<Project> onProjectAnalyzed, CancellationToken cancellationToken)
+            {
+                await _diagnosticAnalyzerService.ForceAnalyzeProjectAsync(project, cancellationToken).ConfigureAwait(false);
+                _analyzedProjectIds.Add(project.Id);
+                onProjectAnalyzed(project);
+                _diagnosticsRefresher.RequestWorkspaceRefresh();
```

Comment:
We invoke a workspace refresh request after each project has been analyzed. This will eventually be replaced with a special workspace refresh request that skips pulling document diagnostics and also does not add any delay for pulling workspace diagnostics.

Reply:
this is fine with me, but just doc :)